app/script/event/EventRepository.js

```diff
@@ -698,27 +698,74 @@ z.event.EventRepository = class EventRepository {
     return this.conversationService.update_event(identifiedUpdates);
   }
 
+  _handleDuplicatedEvent(originalEvent, newEvent) {
+    switch (newEvent.type) {
+      case z.event.Client.CONVERSATION.ASSET_ADD:
+        return this._handleAssetUpdate(originalEvent, newEvent);
+
+      case z.event.Client.CONVERSATION.MESSAGE_ADD:
+        return this._handleLinkPreviewUpdate(originalEvent, newEvent);
+
+      default:
+        this._throwValidationError(newEvent, `Forbidden type '${newEvent.type}' for duplicate events`);
+    }
+  }
+
+  _handleAssetUpdate(originalEvent, newEvent) {
+    const newEventData = newEvent.data;
+    // the preview status is not sent by the client so we fake a 'preview' status in order to cleany handle it in the switch statement
+    const LOCAL_PREVIEW_STATUS = 'preview';
+    const isPreviewEvent = !newEventData.status && newEventData.preview_key;
+    const status = isPreviewEvent ? LOCAL_PREVIEW_STATUS : newEventData.status;
+
+    switch (status) {
+      case LOCAL_PREVIEW_STATUS:
+      case z.assets.AssetTransferState.UPLOADED: {
+        const updatedData = Object.assign({}, originalEvent.data, newEventData);
+        const updatedEvent = Object.assign({}, originalEvent, {data: updatedData});
+        return this.conversationService.update_event(updatedEvent);
+      }
+
+      case z.assets.AssetTransferState.UPLOAD_FAILED: {
+        // case of both failed or canceled upload
+        const fromRemote = newEvent.from !== this.userRepository.self().id;
+        const localCancel = !fromRemote && newEvent.data.reason === z.assets.AssetUploadFailedReason.CANCELLED;
```

Comment:
I am not too fond of the use of remote and local here. As self clients can be remote as well. Let's stick to `self` and `other`